src/queue.h

```diff
@@ -34,75 +37,90 @@ typedef void *QUEUE[2];
-  while (0)
-
-#define QUEUE_MOVE(h, n)                                                      \
-  do {                                                                        \
-    if (QUEUE_EMPTY(h))                                                       \
-      QUEUE_INIT(n);                                                          \
-    else {                                                                    \
-      QUEUE* q = QUEUE_HEAD(h);                                               \
-      QUEUE_SPLIT(h, q, n);                                                   \
-    }                                                                         \
-  }                                                                           \
-  while (0)
-
-#define QUEUE_INSERT_HEAD(h, q)                                               \
-  do {                                                                        \
-    QUEUE_NEXT(q) = QUEUE_NEXT(h);                                            \
-    QUEUE_PREV(q) = (h);                                                      \
-    QUEUE_NEXT_PREV(q) = (q);                                                 \
-    QUEUE_NEXT(h) = (q);                                                      \
-  }                                                                           \
-  while (0)
-
-#define QUEUE_INSERT_TAIL(h, q)                                               \
-  do {                                                                        \
-    QUEUE_NEXT(q) = (h);                                                      \
-    QUEUE_PREV(q) = QUEUE_PREV(h);                                            \
-    QUEUE_PREV_NEXT(q) = (q);                                                 \
-    QUEUE_PREV(h) = (q);                                                      \
-  }                                                                           \
-  while (0)
-
-#define QUEUE_REMOVE(q)                                                       \
-  do {                                                                        \
-    QUEUE_PREV_NEXT(q) = QUEUE_NEXT(q);                                       \
-    QUEUE_NEXT_PREV(q) = QUEUE_PREV(q);                                       \
-  }                                                                           \
-  while (0)
+  for ((q) = (h)->next; (q) != (h); (q) = (q)->next)
+
+/* Removing an element from the queue, while iterating over it,
+ * should be done using QUEUE_REMOVE_SAFE() in conjunction with
+ * QUEUE_FOREACH_SAFE() sharing the same n.
+ * It's still not safe to remove the head though.
+ */
+#define QUEUE_FOREACH_SAFE(q, n, h)                                           \
+  for ((q) = (h)->next, (n) = (q)->next;                                      \
+       (q) != (h);                                                            \
+       (q) = (n), (n) = (n)->next)
+
+static inline int QUEUE_EMPTY(const QUEUE *q) {
```

Comment:
Can you align the pointer stars to the left throughout.

Reply:
Sure thing.
I'll post an updated version after gathering some more comments.